### ISSUE TITLE
Too many versions of requests are blacklisted

### ISSUE THREAD

**Issue by wvxvw** [docker/compose#6593]
Your `setup.py` has this constraint on `requests` library:

``` python
'requests >= 2.6.1, != 2.11.0, != 2.12.2, != 2.18.0, < 2.21'
```

When your package is used with other libraries, it is very difficult (and sometimes impossible) to find a version of another package that would perfectly fit into your constraints.  For example, `prance`, the OpenAPI parser requires `requests ~=2.21`, and there's no way to get a version that wants something that would fit your requirements.

I understand that this requirement must have been a result of bug reports that resulted from using `requests` package, but... please be reasonable, and try to work around those bugs, perhaps by locally patching or even vendoring the version you need.  The way it is now, it effectively prevents the use of your library by projects which need to use other libraries, and `requests` is an almost ubiquitous requirement nowadays...

**Comment by jfinkhaeuser** [comment]
Also, this type of version specification does not follow PEP-0440. Specifially, this is what the `~=` compatible release specification is for.

What you should do is specify the version like this:
```
'requests ~= 2.6.1, != 2.11.0, != 2.12.2, != 2.18.0'
```

This would allow future versions to be included, and you only have maintenance costs with blacklisting known bad versions.

See also jfinkhaeuser/prance#32 .

**Comment by shin-** [comment]
For reference, [this has been discussed before.](https://github.com/docker/compose/issues/4431#issuecomment-277846958)

**Comment by shin-** [comment]
Also, you're misrepresenting [PEP 440](https://www.python.org/dev/peps/pep-0440/#definitions), which doesn't prescribe (or even suggest) how package dependencies should be declared; only how automated tools (like setuptools) should interpret version identifiers.

**Comment by jfinkhaeuser** [comment]
Don't be disingenuous. Of course you can specify version identifiers whichever way you want, but if there's a specific way to solve this particular issue, you have to have very specific reasons not to use them. 

I've read up on those issues you reference, and they don't contradict this usage. You're not going to avoid testing each version either way, but at the very least you default to testing against recent versions, so don't need users to complain for you to upgrade.

**Comment by shin-** [comment]
Disclaimer: I'm no longer a maintainer, so this is all just my personal opinion.

All I'm saying is we've had to make exceptional releases of `docker-py` or Compose on more than one occasion in the past because a new version of `requests` wasn't compatible with the (admittedly unconventional) way we were using it. Those are the versions we've explicitly blacklisted with `!=` clauses; then, because we didn't want this to happen again, we elected to stop allowing versions above a certain point, which stopped the problem from occurring.

Good for you if it's not a problem you've had to deal with, but at the same time I think it's pretty stubborn of you to assume everyone else is dealing with the same problem set.

**Comment by wvxvw** [comment]
@shin- The ticket you refer to was closed w/o any real solution, and the situation has only worsened since.

Just to make sure we are on the same page: if I want to use both `docker-compose` and `prance`, my choices here are:

1. Exploit a bug in PIP I found, and PIP maintainers are refusing to fix (kind of funny... but not a very reliable long term solution). I.e. I can write my `setup.py` in such a way, that it will use `pip` package to install dependencies. This means a lot more work for me, an unnecessary dependency in my `setup.py` and longer times deploying my Python packages (installing PIP on Linux is a circus with elephants and clowns, it drags with itself GCC and a bunch of other unrelated bug huge dependencies).
2. Patch `setuptools` to be able to remove `docker-compose` or `prance` requirements from downloaded archive. I've done this before actually... good times, eh? I will avoid having to rely on bugs in other people's code and my deployment speeds will remain fairly minimal, but the up-front effort is bigger than in case of (1).
3. I can vendor either `prance` or `docker-compose`. I'll probably vendor `prance` if I end up doing this, since `docker-compose` is a much bigger project. This has a tiny fee to pay up-front, but the maintenance cost is huge...

----

Bottom line, I'm out of good or even so-so options. Whatever I choose, is a path of suffering, and none of that suffering was really necessary. I just want you guys to pick up your part of the bill... that's all.

**Comment by jfinkhaeuser** [comment]
@shin- on the contrary, I've been in comparable situations many times, which is why I suggest defaulting to allowing the latest versions. It doesn't stop you from blacklisting explicitly what doesn't work. But it does increase the number of versions that may work without you having to wait for users to raise an issue about it.

Maybe you're explicitly disallowing 2.21 - the commit history suggests, though, that it's simply a case of nobody having gotten around to bumping up the high bound.

**Comment by shin-** [comment]
@wvxvw 
Option 4: Make a PR on Compose to update the dependencies and allow requests 2.21
Option 5: Downgrade prance to 0.13.2

It's not exactly the dire situation you make it seem.

@jfinkhaeuser Allowing the latest versions put us in the situation where we needed to make emergency releases to prevent users from installing with broken dependencies (because they were the latest at the time). Being slightly behind the latest version of `requests` is vastly preferable to that.

**Comment by jfinkhaeuser** [comment]
Your choice :) I would've vastly preferred not to be bothered with upgrade requests. Because users can always choose to lock themselves in to an earlier version than the latest... but fair enough, you need users that know how to do that for that to help.

**Comment by wvxvw** [comment]
@shin- 
(4) Won't work because maybe it's incompatible... and verifying that it is is even more work, which may end up giving a negative result. Anyways, if `docker-compose` makes such a "unique and fragile" use of requests, they should just vendor requests and not interfere with whatever other libraries use. I could make *this kind* of pull request, but what is the chance of it being accepted?

(5) Won't work because previous version uses a version of `requests` that is also blacklisted by `docker-compose`. I don't want t downgrade too many versions, because then I'll miss bug-fixes, meaning, that it will be preferable to me to actually vendor `prance`.

So, those are not just non-solutions, they are non-starters.

**Comment by ulyssessouza** [comment]
Fixed with proposition 4 by https://github.com/docker/compose/pull/6649
